### ISSUE TITLE
Casting real parameter to complex during forward produces warning on backward

### ISSUE THREAD

**Issue by lucgiffon** [pytorch/pytorch#68331]
## 🐛 Bug

If I have a `Parameter` with `real` `dtype` but need to cast it as `complex` in the forward pass, then I get a `UserWarning`  during the backward:

```
/home/luc/anaconda3/envs/ckm/lib/python3.7/site-packages/torch/autograd/__init__.py:156: UserWarning: Casting complex values to real discards the imaginary part (Triggered internally at  /opt/conda/conda-bld/pytorch_1634272092750/work/aten/src/ATen/native/Copy.cpp:244.)
  allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
```

## To Reproduce

```
import torch

def loss(x, w, alpha):
    b = torch.tensor([1 + 1.j], dtype=torch.complex64)
    return (torch.linalg.norm(torch.exp(1.j * torch.tensor([x @ w]) @ alpha) - b)).pow(2)

def main():
    x = torch.nn.Parameter(torch.randn((1,), dtype=torch.float), requires_grad=True)
    alpha = torch.nn.Parameter(torch.tensor([1], dtype=torch.float), requires_grad=True)
    w = torch.tensor([1], requires_grad=False, dtype=torch.float)

    optim = torch.optim.Adam([x], lr=0.1)
    for i in range(1000):
        optim.zero_grad()
        loss_res = loss(x, w, alpha.to(torch.complex64))
        loss_res.backward()
        optim.step()

main()
```

## Expected behavior

I am not sure if the warning should be disabled by default because I can see when there could actually be a mistake and the warning would be desirable. But I think there should be a way to prevent the  warning from occuring as this is the desired behavior.

## Environment

```
Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.2 LTS (beaver-osp1-gilly X33) (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.7.10 (default, Jun  4 2021, 14:48:32)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-89-generic-x86_64-with-debian-buster-sid
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] torch==1.10.0
[pip3] torchaudio==0.10.0
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl  
[conda] cpuonly                   2.0                           0    pytorch
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py37h27cfd23_1  
[conda] mkl_fft                   1.3.0            py37h42c9631_2  
[conda] mkl_random                1.2.1            py37ha9443f7_2  
[conda] numpy                     1.21.1                   pypi_0    pypi
[conda] numpy-base                1.20.1           py37h7d8b39e_0  
[conda] pytorch                   1.10.0              py3.7_cpu_0    pytorch
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torchaudio                0.10.0                 py37_cpu  [cpuonly]  pytorch
[conda] torchvision               0.11.1                 py37_cpu  [cpuonly]  pytorch
```

**Comment by Lezcano** [comment]
This seems like a bug in the implementation of the derivative of `to(dtype)`. That warning should not be happening there. cc @anjali411 

As a side note, remember that you can simply call `.cfloat()` in your case, that has the same semantics as `.to(torch.complex64)` but it's shorter :)

**Comment by Lezcano** [comment]
I don't think we should, as they may be meaningful. In this case, we should see how to deactivate this one in particular.

**Comment by anjali411** [comment]
> I don't think we should, as they may be meaningful.
Mm. can you give an example? I could imagine this being useful for a custom Function, but not otherwise.

**Comment by Lezcano** [comment]
Using a deprecated function in the backwards of a function.

There are not many other examples, as there aren't that many warnings in PyTorch really. In any case, to suppress all the warnings in the backwards we do not need to show that some of them may not be useful, but that all of them are not useful. As such, I would err on the side of precaution and not suppress them.

**Comment by lucgiffon** [comment]
> @lucgiffon thanks for filing the issue! We shouldn't be generating a warning here. Separately, you can easily get rid of the warning by creating an nn parameter alpha by passing in a complex tensor! `alpha = torch.nn.Parameter(torch.tensor([1], dtype=torch.cfloat), requires_grad=True)`

@anjali411 If I am applying the fix you just suggested, wouldn't I risk to end up with an imaginary part different than zero?  In my specific example, it doesn't move from zero but can I be sure it won't ever be the case? The gradient of the imaginary part will always be zero if it is initialized at zero?

**Comment by Lezcano** [comment]
You will discard the imaginary part of the derivative, but that's the derivative of the operation you wrote.
In a similar fashion, if you have a vector `x` of length 1 and you write `y = torch.cat(`x`, torch.zeros(1), dim=0)`, the gradient wrt. `y` is of length 2, but then the gradient wrt to `x` is the first component of the gradient wrt `y`. The same thing is happening here, only that `x` is a real tensor and we are "appending the zeros of the imaginary part to it" when casting it to complex.

**Comment by anjali411** [comment]
Hmm I ran both versions and it seems like the real value of the gradient is not the same in both cases.

**Comment by albanD** [comment]
Having a complex parameter could lead to a non-zero imaginary part indeed.

A simple way to remove the warning is to replace the `param.to(torch.cfloat)` with `param + 0 * 1j` (or even`torch.view_as_complex(torch.stack([param, torch.zeros_like(param)], -1))` if you want to avoid python complex numbers). This will make it clear that the imaginary part is a constant Tensor that doesn't require grad (and so it is ok for the autograd to discard it.

**Comment by Lezcano** [comment]
Even then, I think that we should fix this. I think it's much more readable to write `param.float64()`  or `param.float128()` than the current alternatives.

**Comment by Lezcano** [comment]
As an unrelated sidenote, @lucgiffon, you might want to know that, at the moment, the complex versions of many optimisers do not behave as they should. See https://github.com/pytorch/pytorch/issues/59998 and https://github.com/pytorch/pytorch/issues/65711 for the issue that brought this up and the issue that that tracks this problem respectively.

**Comment by ezyang** [comment]
A bandaid fix is to suppress the warning if the input doesn't require grad (which it won't in typical single backwards). It definitely feels like a hack though. Another thought: maybe we shouldn't emit warnings from code that is not directly called from the user...

**Comment by Lezcano** [comment]
That second point is true almost always, but there are a couple cases that come to mind where this is desirable (e.g. the backward of SVD / eig / eigh, where we throw an error if the function wrt to the eigenvector / singular vectors the user specified is not well-defined. This can just be detected during the backwards pass.

**Comment by Lezcano** [comment]
There's also the usual "the tensor and the gradient don't have the same strides" warning.
